src/lib/hooks/swap/useWrapCallback.tsx

```diff
@@ -0,0 +1,153 @@
+  NO_ERROR, // must be equal to 0 so all other errors are truthy
+  ENTER_NATIVE_AMOUNT,
+  ENTER_WRAPPED_AMOUNT,
+  INSUFFICIENT_NATIVE_BALANCE,
+  INSUFFICIENT_WRAPPED_BALANCE,
+}
+
+export function WrapErrorText({ wrapInputError }: { wrapInputError: WrapInputError }) {
+  const native = useNativeCurrency()
+  const wrapped = native?.wrapped
+
+  switch (wrapInputError) {
+    case WrapInputError.ENTER_NATIVE_AMOUNT:
+      return <Trans>Enter {native?.symbol} amount</Trans>
+    case WrapInputError.ENTER_WRAPPED_AMOUNT:
+      return <Trans>Enter {wrapped?.symbol} amount</Trans>
+    case WrapInputError.INSUFFICIENT_NATIVE_BALANCE:
+      return <Trans>Insufficient {native?.symbol} balance</Trans>
+    case WrapInputError.INSUFFICIENT_WRAPPED_BALANCE:
+      return <Trans>Insufficient {wrapped?.symbol} balance</Trans>
+    case WrapInputError.NO_ERROR:
+    default:
+      return null
+  }
+}
+
+export enum WrapType {
+  NOT_APPLICABLE,
+  WRAP,
+  UNWRAP,
+}
+interface UseWrapCallbackReturns {
+  callback: () => Promise<ContractTransaction>
+  error: WrapInputError
+  loading: boolean
+  type: WrapType
+}
+
+interface WrapState {
+  loading: boolean
+  error: WrapInputError
+}
+
+const wrapState = atom<WrapState>({
+  loading: false,
+  error: WrapInputError.NO_ERROR,
+})
+
+export default function useWrapCallback(): UseWrapCallbackReturns {
+  const { account, chainId } = useActiveWeb3React()
+  const [{ loading, error }, setWrapState] = useAtom(wrapState)
+  const wrappedNativeCurrencyContract = useWETHContract()
+  const {
+    amount,
+    independentField,
+    [Field.INPUT]: inputCurrency,
+    [Field.OUTPUT]: outputCurrency,
+  } = useAtomValue(swapAtom)
+
+  const wrapType = useMemo(() => {
+    if (!inputCurrency || !outputCurrency || !chainId) {
+      return WrapType.NOT_APPLICABLE
+    }
+    const wrappedNativeCurrency = WRAPPED_NATIVE_CURRENCY[chainId]
+    if (inputCurrency.isNative && wrappedNativeCurrency.equals(outputCurrency)) {
+      return WrapType.WRAP
+    }
+    if (wrappedNativeCurrency.equals(inputCurrency) && outputCurrency.isNative) {
+      return WrapType.UNWRAP
+    }
+    return WrapType.NOT_APPLICABLE
+  }, [chainId, inputCurrency, outputCurrency])
+
+  const isExactIn: boolean = independentField === Field.INPUT
+  const parsedAmount = useMemo(
+    () => tryParseCurrencyAmount(amount, (isExactIn ? inputCurrency : outputCurrency) ?? undefined),
+    [inputCurrency, isExactIn, outputCurrency, amount]
+  )
+  const parsedAmountIn = isExactIn ? parsedAmount : undefined
+
+  const relevantTokenBalances = useCurrencyBalances(
+    account,
+    useMemo(() => [inputCurrency ?? undefined, outputCurrency ?? undefined], [inputCurrency, outputCurrency])
+  )
+  const currencyBalances = useMemo(
+    () => ({
+      [Field.INPUT]: relevantTokenBalances[0],
+      [Field.OUTPUT]: relevantTokenBalances[1],
+    }),
+    [relevantTokenBalances]
+  )
+
+  const hasInputAmount = Boolean(parsedAmount?.greaterThan('0'))
```

Comment:
This needs a Boolean wrapper?

Reply:
this value going from undefined -> false -> undefined would trigger a rerun of the `useEffect` below. Not a big deal, but undesirable imo.